src/guest_memory.rs

```diff
@@ -235,23 +238,11 @@ pub trait GuestMemoryRegion: Bytes<MemoryRegionAddress, E = Error> {
 }
 
 /// Represents a container for a collection of GuestMemoryRegion objects.
-///
-/// The main responsibilities of the GuestMemory trait are:
-/// - hide the detail of accessing guest's physical address.
-/// - map a request address to a GuestMemoryRegion object and relay the request to it.
-/// - handle cases where an access request spanning two or more GuestMemoryRegion objects.
-///
-/// Note: all regions in a GuestMemory object must not intersect with each other.
-pub trait GuestMemory {
+/// All regions in a GuestMemoryMut object must not intersect with each other.
+pub trait GuestMemoryMut {
```

Comment:
> both traits allow interior mutability regarding the contents of the guest memory regions, but the difference comes from whether the set of regions can be mutated as well (or so I understand).

Yes, this is correct. Mutability of the contents is kind of implicit in the whole `VolatileMemory` concept, what matters here is the set of regions.

> that can be fixed by making find_region return a guard, or switching back to something like the old do_in_region which is based on closures, so nothing has to be leaked to the outside.

You're not very far from the design that I used here, actually. The guard is the `GuestMemory` returned by a `GuestMemoryMut`'s `snapshot()` method; when the `find_region` method is then invoked on the snapshot itself, the lifetime is limited to that of the snapshot (and therefore the snapshot acts as a guard).

> GuestMemoryMut can only guarantee the same over the course of a single method invocation, but why isn't the latter sufficient?

Basically because it prevents caching. In order to implement full-blown caching, something like a `GuestMemoryMut` observer should be added (also mutuated by QEMU's `MemoryListener` concept), but that would be separate because observers in Rust are tricky.

Reply:
When we are working with Solaris/Linux memory hotplug, we used reference count on pages and observer pattern to reclaim memory on demand. But the whole memory hotplug implementation is really complex, we have struggled with it for years, to fix subsystem/drivers without knowledge about memory hotplug. It's really a hard and long way.

For VMM memory usage patterns, I feel a snapshot mechanism should be a balanced tradeoff between complexity and performance. We only need to maintain region level reference count and address space level snapshots, and changes to GuestMemory consumers to support snapshot should be minor.

For cases which really cares about performance, ArcSwap cache could be used to minimize the   overhead of snapshot, near to zero:)

Reply:
> > that can be fixed by making find_region return a guard, or switching back to something like the old do_in_region which is based on closures, so nothing has to be leaked to the outside.
> 
> You're not very far from the design that I used here, actually. The guard is the `GuestMemory` returned by a `GuestMemoryMut`'s `snapshot()` method; when the `find_region` method is then invoked on the snapshot itself, the lifetime is limited to that of the snapshot (and therefore the snapshot acts as a guard).

Yes, the key idea is to return:
1) a reference for non-hotplug-capable guest memory implement to minimize penalty
2) a guard for hotplug-capable guest memory implementation to ensure correctness.
Thanks to rust's lifetime, it's hard to combine a reference and a guard as one type. Paolo and I are trying to solve the lifetime issue with different methods. For C, it's not an issue at all:(

With Paolo's proposal, a possible pattern for device driver may be something like:
```rust
struct DummyDevice<M: GuestMemoryBackend> {
     vm_memory: M::Memory,
}
DummyDevice<M: GuestMemory> {
      pub fn new(vm_memory: M::Memory {
             DummyDevice {
                        vm_memory
             }
      }
      pub fn handler(&self) {
            let snapshot = self.vm_memory.snapshot();
            self.process_request(snapshot);
      }
      fn process_request(&self, snapshot: M:Snapshot) {
      }
}
```
Basically we need to change existing device driver to this mode, but it should be trivial.

Reply:
Too optimistic, we need to carries on an extra lifetime parameter for GuestMemoryBackend:(

Reply:
Yes, `'a` is still needed but it's much lighter-weight than a generic type parameter, and especially nested generic type parameters which you have to write as `<A, B<A>>`.